File: python/src/prts_mcp/data/story_reader.py
"""Core story data reader — types, constants, and chapter/event parsing.

Split from story.py for module boundaries (see STYLE.md). This module owns
the data types and the low-level store-based reading primitives; higher-level
search, summary, and memoir logic live in their own modules.
"""
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

from prts_mcp.data.stores import JsonStore, ZipStore

# ---------------------------------------------------------------------------
# Zip path constants (public — consumed by sibling story submodules)
# ---------------------------------------------------------------------------

STORY_REVIEW_TABLE = "zh_CN/gamedata/excel/story_review_table.json"
STORYINFO = "zh_CN/storyinfo.json"
SUMMARIES = "zh_CN/summaries.json"
EVENT_SUMMARIES = "zh_CN/event_summaries.json"
CHARDICT = "zh_CN/chardict.json"

# entryType values → user-facing category strings
CATEGORY_MAP: dict[str, list[str]] = {
    "main": ["MAINLINE"],
    "activities": ["ACTIVITY", "MINI_ACTIVITY"],
    "memoirs": ["NONE"],
}


def story_zip_path(story_key: str) -> str:
    return f"zh_CN/gamedata/story/{story_key}.json"


# ---------------------------------------------------------------------------
# Text cleaning
# ---------------------------------------------------------------------------

_RICH_TAG_RE = re.compile(r"<[^>]+>")
_MEMOIR_EVENT_RE = re.compile(r"^story_([a-z0-9_]+)_set_\d+$")


def _clean_text(text: str) -> str:
    """Remove rich-text tags and replace {@nickname} with 博士."""
    text = text.replace("{@nickname}", "博士")
    text = _RICH_TAG_RE.sub("", text)
    return text.strip()


def _is_memoir_event(event_id: str) -> bool:
    """Check if an event_id matches the memoir pattern story_{code}_set_N.

    Public alias ``is_memoir_event`` is provided below for sibling modules.
    """
    return bool(_MEMOIR_EVENT_RE.match(event_id))


def is_memoir_event(event_id: str) -> bool:
    """Public alias for :func:`_is_memoir_event` (used by sibling modules)."""
    return _is_memoir_event(event_id)


# ---------------------------------------------------------------------------
# Data types
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class StoryLine:
    type: Literal["dialog", "narration", "choice"]
    role: str | None   # speaker name; None for narration/choice
    text: str


@dataclass(frozen=True)
class StoryChapter:
    story_key: str
    story_code: str
    story_name: str
    avg_tag: str | None
    event_name: str
    story_info: str
    lines: list[StoryLine]


@dataclass(frozen=True)
class EventInfo:
    event_id: str
    name: str
    entry_type: str
    story_count: int


@dataclass(frozen=True)
class ChapterSummary:
    story_key: str
    story_code: str
    story_name: str
    avg_tag: str | None
    sort_order: int


@dataclass(frozen=True)
class ActivityResult:
    event_id: str
    event_name: str
    total_chapters: int
    has_more: bool
    chapters: list[StoryChapter]


@dataclass(frozen=True)
class MemoirChapter:
    event_id: str
    story_key: str
    story_code: str
    story_name: str


@dataclass(frozen=True)
class OperatorMemoirResult:
    operator_name: str
    internal_code: str
    operator_id: str
    total_chapters: int
    chapters: list[MemoirChapter]


@dataclass(frozen=True)
class CharacterAppearance:
    """A single chapter where a character speaks or is mentioned."""
    event_id: str
    story_key: str
    story_code: str
    story_name: str
    speaks: bool
    mentioned: bool


@dataclass(frozen=True)
class CharacterAppearanceResult:
    """Aggregated appearance report for find_character_appearances."""
    name: str
    total_chapters: int
    appearances: list[CharacterAppearance]


@dataclass(frozen=True)
class SpeakerCount:
    """A speaker and how many dialog lines they have (within one event)."""
    name: str
    line_count: int


# ---------------------------------------------------------------------------
# Store helpers
# ---------------------------------------------------------------------------


def story_store(zip_path: Path) -> ZipStore:
    """Create a ZipStore for the given story zip path (public for sibling modules)."""
    return ZipStore(zip_path)


def load_json(store: JsonStore, path: str) -> dict | list:
    """Read and parse a JSON entry from a store (public for sibling modules)."""
    return store.read_json(path)


# ---------------------------------------------------------------------------
# Core parsing
# ---------------------------------------------------------------------------


def _parse_story_list(story_list: list[dict]) -> list[StoryLine]:
    """Convert raw storyList entries into cleaned StoryLine objects."""
    lines: list[StoryLine] = []
    for item in story_list:
        prop = item.get("prop", "")
        attrs = item.get("attributes", {})

        prop_lower = prop.lower()
        if prop_lower == "name":
            name = attrs.get("name") or ""
            content = attrs.get("content") or ""
            if content:
                lines.append(StoryLine(
                    type="dialog",
                    role=_clean_text(name) if name else None,
                    text=_clean_text(content),
                ))
        elif prop_lower in ("sticker", "subtitle", "animtext"):
            content = attrs.get("content") or attrs.get("text") or ""
            if content:
                lines.append(StoryLine(type="narration", role=None, text=_clean_text(content)))
        elif prop_lower == "decision":
            options = attrs.get("options") or []
            for opt in options:
                # options elements may be plain strings or dicts with a "text" key
                text = opt if isinstance(opt, str) else (opt.get("text") or "")
                if text:
                    lines.append(StoryLine(type="choice", role=None, text=_clean_text(str(text))))

    return lines


# ---------------------------------------------------------------------------
# Public API — zip-path wrappers
# ---------------------------------------------------------------------------


def list_story_events(
    zip_path: Path,
    category: str | None = None,
) -> list[EventInfo]:
    """Return a list of events from story_review_table.json."""
    with story_store(zip_path) as store:
        return list_story_events_from_store(store, category=category)


def list_stories(zip_path: Path, event_id: str) -> list[ChapterSummary]:
    """Return ordered chapter list for an event."""
    with story_store(zip_path) as store:
        return list_stories_from_store(store, event_id)


def read_story(
    zip_path: Path,
    story_key: str,
    include_narration: bool = True,
) -> StoryChapter:
    """Read and parse a single story chapter."""
    with story_store(zip_path) as store:
        return read_story_from_store(store, story_key, include_narration=include_narration)


def read_activity(
    zip_path: Path,
    event_id: str,
    include_narration: bool = True,
    page: int | None = None,
    page_size: int = 5,
) -> ActivityResult:
    """Read all chapters of an activity in official story order."""
    with story_store(zip_path) as store:
        return read_activity_from_store(
            store,
            event_id,
            include_narration=include_narration,
            page=page,
            page_size=page_size,
        )


# ---------------------------------------------------------------------------
# Public API — store-based variants
# ---------------------------------------------------------------------------


def list_story_events_from_store(
    store: JsonStore,
    category: str | None = None,
) -> list[EventInfo]:
    """Return a list of events from story_review_table.json using a JSON store."""
    allowed_types: list[str] | None = CATEGORY_MAP.get(category) if category else None
    table: dict = load_json(store, STORY_REVIEW_TABLE)  # type: ignore[assignment]

    events = []
    for event_id, entry in table.items():
        entry_type = entry.get("entryType", "NONE")
        if allowed_types is not None:
            if allowed_types == ["NONE"]:
                # "memoirs" category: NONE entries whose event_id matches memoir pattern
                if not _is_memoir_event(event_id):
                    continue
            elif entry_type not in allowed_types:
                continue
        story_count = len(entry.get("infoUnlockDatas") or [])
        events.append(EventInfo(
            event_id=event_id,
            name=entry.get("name") or event_id,
            entry_type=entry_type,
            story_count=story_count,
        ))

    return events


def list_stories_from_store(store: JsonStore, event_id: str) -> list[ChapterSummary]:
    """Return ordered chapter list for an event using a JSON store."""
    table: dict = load_json(store, STORY_REVIEW_TABLE)  # type: ignore[assignment]

    entry = table.get(event_id)
    if entry is None:
        raise KeyError(f"Event not found: {event_id!r}")

    chapters = []
    for d in sorted(entry.get("infoUnlockDatas") or [], key=lambda x: x.get("storySort", 0)):
        story_key = d.get("storyTxt")
        if not story_key:
            continue
        chapters.append(ChapterSummary(
            story_key=story_key,
            story_code=d.get("storyCode") or "",
            story_name=d.get("storyName") or "",
            avg_tag=d.get("avgTag"),
            sort_order=d.get("storySort", 0),
        ))

    return chapters


def read_story_from_store(
    store: JsonStore,
    story_key: str,
    include_narration: bool = True,
) -> StoryChapter:
    """Read and parse a single story chapter using a JSON store."""
    story_path = story_zip_path(story_key)
    if not store.exists(story_path):
        raise KeyError(f"Story not found in store: {story_key!r}")
    raw: dict = load_json(store, story_path)  # type: ignore[assignment]

    all_lines = _parse_story_list(raw.get("storyList") or [])
    if not include_narration:
        all_lines = [
            ln for ln in all_lines
            if ln.type != "narration"
            and not (ln.type == "dialog" and ln.role is None)
        ]

    return StoryChapter(
        story_key=story_key,
        story_code=raw.get("storyCode") or "",
        story_name=raw.get("storyName") or "",
        avg_tag=raw.get("avgTag"),
        event_name=raw.get("eventName") or "",
        story_info=raw.get("storyInfo") or "",
        lines=all_lines,
    )


def read_activity_from_store(
    store: JsonStore,
    event_id: str,
    include_narration: bool = True,
    page: int | None = None,
    page_size: int = 5,
) -> ActivityResult:
    """Read all chapters of an activity in official story order using a JSON store."""
    summaries = list_stories_from_store(store, event_id)
    total = len(summaries)

    if page is not None:
        if page < 1:
            raise ValueError("page 参数必须 >= 1")
        start = (page - 1) * page_size
        end = start + page_size
        selected = summaries[start:end]
        has_more = end < total
    else:
        selected = summaries
        has_more = False

    chapters = []
    event_name = ""
    for summary in selected:
        try:
            chapter = read_story_from_store(store, summary.story_key, include_narration)
            if not event_name:
                event_name = chapter.event_name
            chapters.append(chapter)
        except (KeyError, FileNotFoundError, json.JSONDecodeError):
            # Story file missing or corrupt — skip silently
            pass

    return ActivityResult(
        event_id=event_id,
        event_name=event_name,
        total_chapters=total,
        has_more=has_more,
        chapters=chapters,
    )

File: ts/src/data/storyReader.ts
/**
 * Core story data reader — types, constants, and chapter/event parsing.
 *
 * Split from story.ts for module boundaries (see STYLE.md). This module owns
 * the data types and the low-level store-based reading primitives; higher-level
 * search, summary, and memoir logic live in their own modules.
 * Mirrors python/src/prts_mcp/data/story_reader.py.
 */

import { JsonStore, ZipStore } from "./stores.js";

// ---------------------------------------------------------------------------
// Zip path constants
// ---------------------------------------------------------------------------

export const STORY_REVIEW_TABLE = "zh_CN/gamedata/excel/story_review_table.json";
export const STORYINFO = "zh_CN/storyinfo.json";
export const SUMMARIES = "zh_CN/summaries.json";
export const EVENT_SUMMARIES = "zh_CN/event_summaries.json";
export const CHARDICT = "zh_CN/chardict.json";

export function storyZipPath(storyKey: string): string {
  return `zh_CN/gamedata/story/${storyKey}.json`;
}

// entryType → category filter mapping (mirrors Python _CATEGORY_MAP)
export const CATEGORY_MAP: Record<string, string[]> = {
  main: ["MAINLINE"],
  activities: ["ACTIVITY", "MINI_ACTIVITY"],
  memoirs: ["NONE"],
};

// ---------------------------------------------------------------------------
// Text cleaning
// ---------------------------------------------------------------------------

const RICH_TAG_RE = /<[^>]+>/g;
export const MEMOIR_EVENT_RE = /^story_([a-z0-9_]+)_set_\d+$/;

export function cleanText(text: string): string {
  return text.replace(/\{@nickname\}/g, "博士").replace(RICH_TAG_RE, "").trim();
}

export function isMemoirEvent(eventId: string): boolean {
  return MEMOIR_EVENT_RE.test(eventId);
}

// ---------------------------------------------------------------------------
// JSON shape types (only fields we use)
// ---------------------------------------------------------------------------

interface RawStoryItem {
  prop?: string;
  attributes?: Record<string, unknown>;
}

interface RawStoryChapter {
  storyCode?: string;
  storyName?: string;
  avgTag?: string | null;
  eventName?: string;
  storyInfo?: string;
  storyList?: RawStoryItem[];
}

interface RawInfoUnlockData {
  storyTxt?: string;
  storyCode?: string;
  storyName?: string;
  avgTag?: string | null;
  storySort?: number;
}

interface RawReviewEntry {
  id?: string;
  name?: string;
  entryType?: string;
  infoUnlockDatas?: RawInfoUnlockData[];
}

export type RawReviewTable = Record<string, RawReviewEntry>;

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

export interface StoryLine {
  type: "dialog" | "narration" | "choice";
  role: string | null;
  text: string;
}

export interface StoryChapter {
  storyKey: string;
  storyCode: string;
  storyName: string;
  avgTag: string | null;
  eventName: string;
  storyInfo: string;
  lines: StoryLine[];
}

export interface EventInfo {
  eventId: string;
  name: string;
  entryType: string;
  storyCount: number;
}

export interface ChapterSummary {
  storyKey: string;
  storyCode: string;
  storyName: string;
  avgTag: string | null;
  sortOrder: number;
}

export interface ActivityResult {
  eventId: string;
  eventName: string;
  totalChapters: number;
  hasMore: boolean;
  chapters: StoryChapter[];
}

export interface MemoirChapter {
  eventId: string;
  storyKey: string;
  storyCode: string;
  storyName: string;
}

export interface OperatorMemoirResult {
  operatorName: string;
  internalCode: string;
  operatorId: string;
  totalChapters: number;
  chapters: MemoirChapter[];
}

export interface CharacterAppearance {
  /** A single chapter where a character speaks or is mentioned. */
  eventId: string;
  storyKey: string;
  storyCode: string;
  storyName: string;
  speaks: boolean;
  mentioned: boolean;
}

export interface CharacterAppearanceResult {
  /** Aggregated appearance report for findCharacterAppearances. */
  name: string;
  totalChapters: number;
  appearances: CharacterAppearance[];
}

export interface SpeakerCount {
  /** A speaker and how many dialog lines they have (within one event). */
  name: string;
  lineCount: number;
}

export interface StoryEventsListingPayload {
  total: number;
  filters: {
    category: string | null;
    category_normalized: string | null;
  };
  events: Array<{
    event_id: string;
    name: string;
    entry_type: string;
    story_count: number;
  }>;
}

export interface StoriesListingPayload {
  event_id: string;
  total: number;
  include_summaries: boolean;
  chapters: Array<{
    story_code: string;
    story_name: string;
    story_key: string;
    avg_tag: string | null;
    summary?: string;
  }>;
  event_summary?: string;
}

// ---------------------------------------------------------------------------
// Store helpers
// ---------------------------------------------------------------------------

export function storyStore(zipPath: string): ZipStore {
  return new ZipStore(zipPath);
}

export function withStoryStore<T>(zipPath: string, read: (store: ZipStore) => T): T {
  const store = storyStore(zipPath);
  try {
    return read(store);
  } finally {
    store.close();
  }
}

function parseStoryList(
  storyList: RawStoryItem[],
  includeNarration: boolean
): StoryLine[] {
  const lines: StoryLine[] = [];
  for (const item of storyList) {
    const prop = (item.prop ?? "").toLowerCase();
    const attrs = (item.attributes ?? {}) as Record<string, unknown>;

    if (prop === "name") {
      const name = String(attrs["name"] ?? "");
      const content = String(attrs["content"] ?? "");
      if (content && (includeNarration || name)) {
        lines.push({
          type: "dialog",
          role: name ? cleanText(name) : null,
          text: cleanText(content),
        });
      }
    } else if (
      includeNarration &&
      (prop === "sticker" || prop === "subtitle" || prop === "animtext")
    ) {
      // Sticker uses "text" OR "content" field — check both (see porting guide §4.2)
      const content = String(attrs["content"] ?? attrs["text"] ?? "");
      if (content) {
        lines.push({ type: "narration", role: null, text: cleanText(content) });
      }
    } else if (prop === "decision") {
      const options = attrs["options"];
      if (Array.isArray(options)) {
        for (const opt of options) {
          let text = "";
          if (typeof opt === "string") text = opt;
          else if (opt !== null && typeof opt === "object") {
            text = String((opt as Record<string, unknown>)["text"] ?? "");
          }
          if (text) {
            lines.push({ type: "choice", role: null, text: cleanText(text) });
          }
        }
      }
    }
    // All other props (Dialog, Background, PlayMusic, etc.) are skipped.
  }
  return lines;
}

// ---------------------------------------------------------------------------
// Public API — zip-path wrappers
// ---------------------------------------------------------------------------

export function listStoryEvents(
  zipPath: string,
  category?: string
): EventInfo[] {
  return withStoryStore(zipPath, (store) => listStoryEventsFromStore(store, category));
}

export function buildStoryEventsListing(
  zipPath: string,
  category?: string,
): StoryEventsListingPayload {
  return withStoryStore(zipPath, (store) => buildStoryEventsListingFromStore(store, category));
}

export function pyRepr(value: string | null | undefined): string {
  if (value === null || value === undefined) return "None";
  const quote = value.includes("'") && !value.includes('"') ? '"' : "'";
  let out = quote;
  for (const ch of value) {
    if (ch === "\\") out += "\\\\";
    else if (ch === "\n") out += "\\n";
    else if (ch === "\r") out += "\\r";
    else if (ch === "\t") out += "\\t";
    else if (ch === quote) out += `\\${quote}`;
    else {
      const code = ch.codePointAt(0) ?? 0;
      out += code < 32 || code === 127
        ? `\\x${code.toString(16).padStart(2, "0")}`
        : ch;
    }
  }
  return out + quote;
}

export function renderStoryEventsListing(data: StoryEventsListingPayload): string {
  if (data.events.length === 0) {
    return `未找到符合条件的活动（category=${pyRepr(data.filters.category)}）。`;
  }

  return data.events
    .map((ev) => `- [${ev.entry_type}] ${ev.event_id}：${ev.name}（${ev.story_count} 章）`)
    .join("\n");
}

export function listStories(
  zipPath: string,
  eventId: string
): ChapterSummary[] {
  return withStoryStore(zipPath, (store) => listStoriesFromStore(store, eventId));
}

export function buildStoriesListing(
  zipPath: string,
  eventId: string,
  includeSummaries = false,
): StoriesListingPayload {
  return withStoryStore(zipPath, (store) =>
    buildStoriesListingFromStore(store, eventId, includeSummaries),
  );
}

export function renderStoriesListing(data: StoriesListingPayload): string {
  if (data.chapters.length === 0) {
    return `活动 ${pyRepr(data.event_id)} 暂无剧情章节。`;
  }

  const lines: string[] = [];
  if (data.event_summary) {
    lines.push(data.event_summary, "");
  }
  for (const ch of data.chapters) {
    const tag = ch.avg_tag ? `[${ch.avg_tag}] ` : "";
    lines.push(`- ${ch.story_code} ${tag}${ch.story_name}（key: ${ch.story_key}）`);
    if (data.include_summaries && ch.summary) {
      lines.push(`  ${ch.summary}`);
    }
  }
  return lines.join("\n");
}

export function readStory(
  zipPath: string,
  storyKey: string,
  includeNarration = true
): StoryChapter {
  return withStoryStore(zipPath, (store) => readStoryFromStore(store, storyKey, includeNarration));
}

export function readActivity(
  zipPath: string,
  eventId: string,
  includeNarration = true,
  page?: number,
  pageSize = 5
): ActivityResult {
  return withStoryStore(zipPath, (store) => readActivityFromStore(
    store,
    eventId,
    includeNarration,
    page,
    pageSize,
  ));
}

// ---------------------------------------------------------------------------
// Public API — store-based variants
// ---------------------------------------------------------------------------

export function listStoryEventsFromStore(
  store: JsonStore,
  category?: string
): EventInfo[] {
  const allowedTypes = category ? (CATEGORY_MAP[category] ?? null) : null;
  const table = store.readJson<RawReviewTable>(STORY_REVIEW_TABLE);

  const events: EventInfo[] = [];
  for (const [eventId, entry] of Object.entries(table)) {
    const entryType = entry.entryType ?? "NONE";
    if (allowedTypes !== null) {
      if (allowedTypes.length === 1 && allowedTypes[0] === "NONE") {
        // "memoirs" category: NONE entries whose eventId matches memoir pattern
        if (!isMemoirEvent(eventId)) continue;
      } else if (!allowedTypes.includes(entryType)) {
        continue;
      }
    }
    events.push({
      eventId,
      name: entry.name ?? eventId,
      entryType,
      storyCount: (entry.infoUnlockDatas ?? []).length,
    });
  }
  return events;
}

export function buildStoryEventsListingFromStore(
  store: JsonStore,
  category?: string,
): StoryEventsListingPayload {
  const events = listStoryEventsFromStore(store, category);
  const categoryNormalized = category && category in CATEGORY_MAP ? category : null;
  return {
    total: events.length,
    filters: {
      category: category ?? null,
      category_normalized: categoryNormalized,
    },
    events: events.map((ev) => ({
      event_id: ev.eventId,
      name: ev.name,
      entry_type: ev.entryType,
      story_count: ev.storyCount,
    })),
  };
}

export function listStoriesFromStore(
  store: JsonStore,
  eventId: string
): ChapterSummary[] {
  const table = store.readJson<RawReviewTable>(STORY_REVIEW_TABLE);

  const entry = table[eventId];
  if (!entry) throw new Error(`Event not found: "${eventId}"`);

  const datas = (entry.infoUnlockDatas ?? []).slice();
  datas.sort((a, b) => (a.storySort ?? 0) - (b.storySort ?? 0));

  const chapters: ChapterSummary[] = [];
  for (const d of datas) {
    if (!d.storyTxt) continue;
    chapters.push({
      storyKey: d.storyTxt,
      storyCode: d.storyCode ?? "",
      storyName: d.storyName ?? "",
      avgTag: d.avgTag ?? null,
      sortOrder: d.storySort ?? 0,
    });
  }
  return chapters;
}

export function buildStoriesListingFromStore(
  store: JsonStore,
  eventId: string,
  includeSummaries = false,
): StoriesListingPayload {
  const chapters = listStoriesFromStore(store, eventId);
  let chapterSummaries: Record<string, string> = {};
  let eventSummaryText = "";

  if (includeSummaries) {
    try {
      if (store.exists(STORYINFO)) {
        const raw = store.readJson<Record<string, unknown>>(STORYINFO);
        chapterSummaries = Object.fromEntries(
          Object.entries(raw)
            .filter(([, value]) => Boolean(value))
            .map(([key, value]) => [key, String(value)]),
        );
      }
      if (store.exists(EVENT_SUMMARIES)) {
        const rawEvents = store.readJson<Record<string, unknown>>(EVENT_SUMMARIES);
        const text = rawEvents[eventId];
        if (typeof text === "string") eventSummaryText = text.trim();
      }
    } catch {
      chapterSummaries = {};
      eventSummaryText = "";
    }
  }

  const data: StoriesListingPayload = {
    event_id: eventId,
    total: chapters.length,
    include_summaries: includeSummaries,
    chapters: chapters.map((ch) => ({
      story_code: ch.storyCode,
      story_name: ch.storyName,
      story_key: ch.storyKey,
      avg_tag: ch.avgTag,
      ...(includeSummaries ? { summary: chapterSummaries[ch.storyKey] ?? "" } : {}),
    })),
  };
  if (eventSummaryText) data.event_summary = eventSummaryText;
  return data;
}

export function readStoryFromStore(
  store: JsonStore,
  storyKey: string,
  includeNarration = true
): StoryChapter {
  const innerPath = storyZipPath(storyKey);
  if (!store.exists(innerPath)) {
    throw new Error(`Story not found in store: "${storyKey}"`);
  }
  const raw = store.readJson<RawStoryChapter>(innerPath);

  const lines = parseStoryList(raw.storyList ?? [], includeNarration);

  return {
    storyKey,
    storyCode: raw.storyCode ?? "",
    storyName: raw.storyName ?? "",
    avgTag: raw.avgTag ?? null,
    eventName: raw.eventName ?? "",
    storyInfo: raw.storyInfo ?? "",
    lines,
  };
}

export function readActivityFromStore(
  store: JsonStore,
  eventId: string,
  includeNarration = true,
  page?: number,
  pageSize = 5
): ActivityResult {
  const summaries = listStoriesFromStore(store, eventId);
  const total = summaries.length;

  let selected: ChapterSummary[];
  let hasMore: boolean;
  if (page !== undefined) {
    if (page < 1) throw new Error("page 参数必须 >= 1");
    const start = (page - 1) * pageSize;
    const end = start + pageSize;
    selected = summaries.slice(start, end);
    hasMore = end < total;
  } else {
    selected = summaries;
    hasMore = false;
  }

  const chapters: StoryChapter[] = [];
  let eventName = "";
  for (const summary of selected) {
    try {
      const chapter = readStoryFromStore(store, summary.storyKey, includeNarration);
      if (!eventName) eventName = chapter.eventName;
      chapters.push(chapter);
    } catch (err) {
      if (err instanceof SyntaxError) { /* corrupt JSON */ }
      else if (err instanceof Error && /not found/i.test(err.message)) { /* missing */ }
      else throw err;
    }
  }

  return { eventId, eventName, totalChapters: total, hasMore, chapters };
}
